### PR TITLE
Duplicate gene name handling in DE processing (SCP-4283)

### PR DESCRIPTION
**Pull request by jlchang**
SCP allows mtx feature files to have duplicate gene names if the accompanying gene ID information uniquifies the feature name. Inputting duplicate gene names into DE analysis results in an anomalous number of entries for the gene. To avoid confusing output, when duplicate gene names are detected, a merged gene name and gene id is used as the input gene name and the name is subsequently split out (to generate an additional gene_id column) in the DE output file.

activate the scp-ingest-pipeline repo virtualenv
then from the scripts directory of the scp-ingest-pipeline repo, perform this setup:

```
source ../scripts/setup_mongo_dev.sh <path to your Github token file>
unset BARD_HOST_URL
```

Run the DE job from the ingest directory of the scp-ingest-pipeline repo:
`python ingest_pipeline.py --study-id addedfeed000000000000000 --study-file-id dec0dedfeed1111111111111 differential_expression --annotation-name cell_type__ontology_label --annotation-type group --annotation-scope study --matrix-file-path ../tests/data/differential_expression/sparse/sparsemini_matrix.mtx.gz --gene-file ../tests/data/differential_expression/sparse/sparsemini_dup_gene_name.tsv --barcode-file ../tests/data/differential_expression/sparse/sparsemini_barcodes.tsv.gz --matrix-file-type mtx --annotation-file ../tests/data/differential_expression/sparse/sparsemini_metadata.txt --cluster-file ../tests/data/differential_expression/sparse/sparsemini_cluster.txt --cluster-name de_sparse_dup_gene --study-accession SCPsparsedup --differential-expression `

confirm that the job runs successfully and the output files have both a "names" column (where "Sox17" occurs twice) and a "gene_id" column 

> de_sparse_dup_gene--cell_type__ontology_label--endothelial_cell--study--wilcoxon.tsv
> de_sparse_dup_gene--cell_type__ontology_label--endothelial_cell_of_lymphatic_vessel--study--wilcoxon.tsv
> de_sparse_dup_gene--cell_type__ontology_label--fibroblast--study--wilcoxon.tsv
> de_sparse_dup_gene--cell_type__ontology_label--glial_cell--study--wilcoxon.tsv
> de_sparse_dup_gene--cell_type__ontology_label--interstitial_cell_of_Cajal--study--wilcoxon.tsv
> de_sparse_dup_gene--cell_type__ontology_label--pericyte_cell--study--wilcoxon.tsv
> de_sparse_dup_gene--cell_type__ontology_label--smooth_muscle_cell--study--wilcoxon.tsv

This PR satisfies SCP-4283.

### CODE REVIEW

**Comment by jlchang** [comment]
Test failures are due EBI OLS lookup failures for known good metadata. I've put in an inquiry about the status of their server https://github.com/EBISPOT/OLS/issues/602

File: ingest/de.py [hunk 1]
```diff
@@ -225,27 +225,39 @@ def execute_de(self):
     def get_genes(genes_path):
         """ Genes file can have one or two columns of gene information
             If two columns present, check if there are duplicates in 2nd col
-            If no duplicates, use as var_names, else use 1st column
+            If no duplicates, use as var_names, else use join columns
```
Comment (eweitz) [inline ingest/de.py:228]:
```suggestion
            If no duplicates, use as var_names, else use joined columns
```

File: tests/test_de.py [hunk 2]
```diff
@@ -89,6 +89,27 @@ def test_assess_annotation(self):
             KeyError, DifferentialExpression.assess_annotation, test_annotation, cm
         )
 
+    def test_detect_duplicate_gene_names(self):
+        """Genes file can have one or two columns of gene information
+            If two columns present, use the second column containing gene names
+            unless there are duplicate gene names in the second column
+            If duplicates, check that 1st plus 2nd column provides uniqueness
+            If unique when joined, join columns with pipe (|) for use as DE input
+        """
```
Comment (eweitz) [inline tests/test_de.py:98]: This is a really helpful summary!

**Review by eweitz — APPROVED** [review]
Looks good!

**Review by bistline — APPROVED** [review]
Looks good - I wasn't able to get the script to run initially because there were typos in the command, and all I got was an exit status of `1` w/o any other info.  The cause of the problem was that the matrix/genes/features files all had a suffix of `.gz` in the example, but those files don't exist.  Trimming off the `.gz` allowed everything to run as specified.

**Pull request merged**